type.go

```diff
@@ -62,6 +62,12 @@ type Type interface {
 	// For other types, the value is zero.
 	Length() int
 
+	// Returns an estimation of the number of bytes required to hold the given
+	// number of values of this type in memory.
+	//
+	// The method returns zero for group types.
+	EstimateSize(numValues int) int64
```

Comment:
Can we be more precise about the estimate? Is it always an exact estimate OR overestimate, or is it always an exact estimate OR underestimate, or, neither

Reply:
Sometimes it overestimates, sometimes it might underestimate, and sometimes it is an exact count.

For example, with boolean columns it always over-estimate since it rounds up to the nearest byte (the values are bit-packed).

For fixed-size types like int32, int64, etc... it's always the exact count.

For variable size types (BYTE_ARRAY) it's an approximation which may either underestimate or overestimate because we can't predict the length of values that the application will be working with.

I'm thinking there are risks with being more precise, users may try to be too smart for their own good by relying on the documented behavior, it also gives us less flexibility to change the behavior later on (if it is undocumented we have a stronger argument for "you should not have taken a dependency on this behavior").

I'm tempted to say that in this case we are better off making few promises, the value returned is a useful hint but users are not expected to rely on it beyond that.

What do you think?

Reply:
Yeah that's OK with me.